pitop/processing/algorithms/ball_detect.py

```diff
@@ -0,0 +1,269 @@
+        self._print_fps = getenv("PT_ENABLE_FPS", "0") == "1"
+        if self._print_fps:
+            self._fps = FPS().start()
+            atexit.register(self.__print_fps)
+
+    def detect(self, frame, color: Union[str, list] = "red"):
+        def parse_colors(color_arg):
+            colors = []
+            if type(color_arg) == str:
+                assert(color_arg in valid_colors)
+                colors = [color_arg]
+            elif type(color_arg) in (list, tuple):
+                assert(set(color_arg).issubset(valid_colors))
+                colors = color_arg
+            if len(colors) > 3:
+                raise ValueError("Cannot pass more than three colors.")
+
+            return colors
+
+        frame = ImageFunctions.convert(frame, format="OpenCV")
+
+        if self._frame_scaler is None:
+            _, width = frame.shape[0:2]
+            self._frame_scaler = width / self._process_image_width
+
+        for c in parse_colors(color):
+            self.balls[c] = self.__find_most_likely_ball(ball=self.balls.get(c),
+                                                         color=c,
+                                                         frame=frame)
+
+        robot_view = frame.copy()
+        ball_data = DotDict({})
+        for ball_color, ball_object in self.balls.items():
+            ball_data[ball_color] = ball_object
+            if ball_object.is_valid():
+                self.__draw_ball_position(robot_view, ball_object)
+                self.__draw_ball_contrail(robot_view, ball_object)
+
+        ball_data["robot_view"] = ImageFunctions.convert(robot_view, self.format)
+
+        if self._print_fps:
+            self._fps.update()
+
+        return ball_data
+
+    def __print_fps(self):
+        self._fps.stop()
+        print(f"[INFO] Elapsed time: {self._fps.elapsed():.2f}")
+        print(f"[INFO] Approx. FPS: {self._fps.fps():.2f}")
+
+    def __draw_ball_position(self, frame, ball):
+        self.cv2.circle(frame, ball.center, ball.radius, (0, 255, 255), 2)
+        self.cv2.circle(frame, ball.center, 5, tuple_for_color_by_name(ball.color, bgr=True), -1)
+
+    def __draw_ball_contrail(self, frame, ball):
+        for i in range(1, len(ball.center_points)):
+            if ball.center_points[i - 1] is None or ball.center_points[i] is None:
+                continue
+            thickness = int(np.sqrt(DETECTION_POINTS_BUFFER_LENGTH / float(i + 1)))
+
+            self.cv2.line(frame, ball.center_points[i - 1], ball.center_points[i],
+                          tuple_for_color_by_name(ball.color, bgr=True), thickness)
+
+    def color_filter(self,
+                     frame,
+                     color: str = "red",
+                     return_binary_mask: bool = False,
+                     format: str = "PIL"
+                     ):
+
+        frame = ImageFunctions.convert(frame, format="OpenCV")
+
+        if color.lower() not in ("red", "green", "blue"):
```

Comment:
This is not _reaaaally_ what a tuple is for - (r, g, b), (x, y) and the like. This would be better as a list. Also, we check for this with `valid_colors` at the top - we might want to create a function to do this check as we have 2 places that use it.